// iotex-antenna/lib/action/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionError = exports.ActionErrorCode = void 0;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

let ActionErrorCode;
exports.ActionErrorCode = ActionErrorCode;

(function (ActionErrorCode) {
  ActionErrorCode[ActionErrorCode["ErrExistedAction"] = 0] = "ErrExistedAction";
  ActionErrorCode[ActionErrorCode["ErrBalance"] = 1] = "ErrBalance";
  ActionErrorCode[ActionErrorCode["ErrNonce"] = 2] = "ErrNonce";
  ActionErrorCode[ActionErrorCode["ErrAddress"] = 3] = "ErrAddress";
  ActionErrorCode[ActionErrorCode["ErrGasPrice"] = 4] = "ErrGasPrice";
  ActionErrorCode[ActionErrorCode["ErrUnknown"] = 5] = "ErrUnknown";
})(ActionErrorCode || (exports.ActionErrorCode = ActionErrorCode = {}));

class ActionError extends Error {
  constructor(code, message) {
    super(message);

    _defineProperty(this, "code", void 0);

    this.code = code;
    Object.setPrototypeOf(this, ActionError.prototype);
  }

}

exports.ActionError = ActionError;
